### PR TITLE
Perform garbage collection for efb copies

### PR DESCRIPTION
**Pull request by mimimi085181**
This checks every TEXTURE_KILL_THRESHOLD frames, to see if the hash for the memory area of the efb copy has hanged. If it has changed, the efb copy can be removed, it wouldn't be used anymore. Before this pr, some efb copies would never be deleted.

Fixes issue https://bugs.dolphin-emu.org/issues/6101 and possibly some other VRAM leaks.

### CODE REVIEW

**Comment by phire** [comment]
The only problem I see is that this may negatively effect NSMB: Wii
(I'm not sure, I still don't fully understand what NSMB; Wii does)

Might want to add an exception for efb copies where stride != width, or (better solution) implement strided hashes.

**Comment by mimimi085181** [comment]
New Super Mario Bros does not need an exception, because it uses the efb copies directly at the same frame. And this cleanup takes place 60 frames after the last use or creation(when not "used"), so it's safe.

Strided hashes is something that i might implement, but that does not belong into this pr. There's only NSMB known to use them anyways.

**Comment by phire** [comment]
True

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 1]
```diff
@@ -151,9 +151,10 @@ void TextureCache::Cleanup(int _frameCount)
 		{
 			iter->second->frameCount = _frameCount;
 		}
-		if (_frameCount > TEXTURE_KILL_THRESHOLD + iter->second->frameCount &&
-		    // EFB copies living on the host GPU are unrecoverable and thus shouldn't be deleted
-		    !iter->second->IsEfbCopy())
+		// Only remove EFB copies when they wouldn't be used anymore(changed hash), because EFB copies living on the
+		// host GPU are unrecoverable. Perform this check only every TEXTURE_KILL_THRESHOLD for performance reasons
+		if ((_frameCount - iter->second->frameCount) % TEXTURE_KILL_THRESHOLD == 1 && (!iter->second->IsEfbCopy() ||
```
Comment (degasus) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: Sounds like a bit too complicated logic. What do you think about:
```
if (iter->second->IsEfbCopy() && iter->second->hash == GetHash64(Memory::GetPointer(iter->second->addr))
    iter->second->frameCount = _frameCount;
```

Comment (mimimi085181) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: This way the frameCount for the cache entry would not contain the info when it was last used anymore. Like some other parts of the code check which entry can be removed, by choosing the oldest one. With this, the most likely oldest one, would become a fairly new entry. Also, who knows what this info could be used in the future.

But the logic is a bit more complicated than i would like it, yeah.

Comment (phire) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: Ok, but perhaps we could split this if statement up a bit, put the hash checking logic inside the if statement. Short-circuit evaluation can be a bit harder to understand at times.

    if (_frameCount - iter->second->frameCount) % TEXTURE_KILL_THRESHOLD == 1)
    {
        if (iter->second->IsEfbCopy())
        {
            if (iter->second->hash != GetHash64(.....))
                iter = FreeTexture(iter);
            else
                ++iter;
        }
        else
        {
               iter = FreeTexture(iter);
        }
    }

Comment (phire) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: Also, now that I've split up the logic, wouldn't this result in non-efb copies being freed the frame after they were created? 

    ((X+1) - X) mod Y == 1

**Comment by phire** [comment]
LGTM.

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 2]
```diff
@@ -150,16 +150,35 @@ void TextureCache::Cleanup(int _frameCount)
...
-		    // EFB copies living on the host GPU are unrecoverable and thus shouldn't be deleted
-		    !iter->second->IsEfbCopy())
-		{
-			iter = FreeTexture(iter);
+			++iter;
 		}
 		else
 		{
-			++iter;
+			if (_frameCount > TEXTURE_KILL_THRESHOLD + iter->second->frameCount)
```
Comment (degasus) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: No need to indent as much, just use "else if()"

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 3]
```diff
@@ -150,16 +150,35 @@ void TextureCache::Cleanup(int _frameCount)
...
+			{
+				if (iter->second->IsEfbCopy())
+				{
+					// Only remove EFB copies when they wouldn't be used anymore(changed hash), because EFB copies living on the
+					// host GPU are unrecoverable. Perform this check only every TEXTURE_KILL_THRESHOLD for performance reasons
+					if ((_frameCount - iter->second->frameCount) % TEXTURE_KILL_THRESHOLD == 1 &&
```
Comment (degasus) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: Are you sure about the order?
_frameCount increase by one per frame. So a frame later, this difference is 1.

"iter->second->frameCount - _frameCount"

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 4]
```diff
@@ -150,12 +150,28 @@ void TextureCache::Cleanup(int _frameCount)
 		if (iter->second->frameCount == FRAMECOUNT_INVALID)
 		{
 			iter->second->frameCount = _frameCount;
+			++iter;
 		}
-		if (_frameCount > TEXTURE_KILL_THRESHOLD + iter->second->frameCount &&
-		    // EFB copies living on the host GPU are unrecoverable and thus shouldn't be deleted
-		    !iter->second->IsEfbCopy())
+		else if (_frameCount > TEXTURE_KILL_THRESHOLD + iter->second->frameCount)
```
Comment (degasus) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: Maybe:
```
int age = _frameCount - iter->second->frameCount;
if (age > TEXTURE_KILL_THRESHOLD)
{
   ...
   if (age % TEXTURE_KILL_THRESHOLD == 1 && something_about_hash)
   {
        ...
```

But it LGTM also right now.

**Comment by Kodiack** [comment]
Awesome PR! It's nice to finally see that bug squashed. Would it have affected this as well?

https://bugs.dolphin-emu.org/issues/7554

I'm no longer able to reproduce that issue, short of doing stupid things. Pokémon Puzzle League just sits at a comfortable, constant commit size at the main menu. If you bump the resolution up/down, it'll spike for a moment, and then almost immediately fall into a stable position.

The only way that I can make the problem happen is if I constantly switch between IRs before things can be cleaned up. I managed to get Dolphin up to ~18 GB before my system locked up for a second, then Dolphin fell down to the otherwise expected ~2-3.5 GB range, and wouldn't budge unless I changed IR again.

**Comment by phire** [comment]
Hmm, is that 18gb of pure EFB copies?

**Comment by phire** [comment]
But yeah. This should fix that bug, if #2960 didn't already. Every texture should be deleted if it's not used within 60 frames, though efb copies only get evicted if there is zero possibility of them being used in the future (hash check fails)

This and #2960 are a pair, attacking two different aspects of this problem.

**Pull request merged**